Fix #9134 - Update pgx to v4

ZPascal opened the pull request:
- Update pgx to v4
- Update go.mod
- Add license information
- Update PgBouncer test

### Required for all PRs:

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

This change fixes a bug inside the pgx library. Until now, it was unfortunately not possible to validate a CA certificate in verify-ca mode. The update of the library to V4 solved the corresponding problem. More details can be found, in the bug linked below.

resolves #9134

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

example: Fixed an off by one error in counter variable in type FooBar.

example: Added an input plugin to gather yak shaving metrics using
golang library yaktech/shaver. -->

I updated pgx to version v4.6.0, fixed a typo, adapted the corresponding documentation and the PgBouncer functionality inside the `service.go` file to the new library version.

@helenosheaa Could you please, review this change?

sspaink commented:
!signed-cla

Jagularr reviewed (approved):
Makes sense to me

helenosheaa commented on plugins/inputs/postgresql/service.go:
```diff
@@ -111,33 +110,14 @@ func (p *Service) Start(telegraf.Accumulator) (err error) {
-				RuntimeParams: map[string]string{
-					"client_encoding": "UTF8",
-				},
-				CustomConnInfo: func(c *pgx.Conn) (*pgtype.ConnInfo, error) {
```
Do we no longer need any driver config? 

The [comment on 111 ](https://github.com/influxdata/telegraf/issues/3253#issuecomment-357505343) points to a comment around this.

Also [this thread](https://github.com/jackc/pgx/issues/649#issuecomment-585547940) seems to suggest we need/needed to register data types?

ZPascal replied:
In my opinion, the configuration of the driver is no longer needed inside v4. However, I'm not 100% sure and I don't know enough about PGBounder. For this reason, I will first try to create an integration test and clear the corresponding doubts based on this test.

helenosheaa replied:
That's great thanks

helenosheaa commented on scripts/check-deps.sh:
```diff
@@ -28,9 +28,6 @@ for dep in $(LC_ALL=C sort -u "${tmpdir}/golist"); do
 		# ignore ourselves
 		github.com/influxdata/telegraf) continue;;
 
-		# dependency is replaced in go.mod
```
Related to https://github.com/satori/go.uuid/issues/73, is this resolved?

ZPascal replied:
> Were you able to run the integration tests? As we don't currently run them as part of our circleci checks.

I ran the appropriate integration tests locally and looking especially at the tests where I made changes to the actual plugins.

ZPascal replied:
The problem should have been fixed [here](https://github.com/satori/go.uuid/pull/75). However when I looked at the repository a second time, I noticed that it's not really maintained. For this reason, I think it is better to implement the replace function.

helenosheaa reviewed (commented):
Thanks for the PR! Overall looks good just a few questions in the code. 

Were you able to run the integration tests? As we don't currently run them as part of our circleci checks.

ZPascal commented:
@helenosheaa 
> Were you able to run the integration tests? As we don't currently run them as part of our circleci checks.

I executed the appropriate integration tests locally and looking especially at the tests where I made changes to the actual plugins.

ZPascal commented on go.mod:
```diff
@@ -148,6 +146,3 @@ require (
 	modernc.org/sqlite v1.7.4
 )
-
-// replaced due to https://github.com/satori/go.uuid/issues/73
```
The problem should have been fixed, but I think it's the better solution to use the replacement functionality.

reimda commented on plugins/inputs/postgresql/service.go:
```diff
@@ -111,33 +110,14 @@ func (p *Service) Start(telegraf.Accumulator) (err error) {
 	// Specific support to make it work with PgBouncer too
 	// See https://github.com/influxdata/telegraf/issues/3253#issuecomment-357505343
 	if p.IsPgBouncer {
-		d := &stdlib.DriverConfig{
-			ConnConfig: pgx.ConnConfig{
-				PreferSimpleProtocol: true,
-				RuntimeParams: map[string]string{
-					"client_encoding": "UTF8",
-				},
-				CustomConnInfo: func(c *pgx.Conn) (*pgtype.ConnInfo, error) {
-					info := c.ConnInfo.DeepCopy()
-					info.RegisterDataType(pgtype.DataType{
-						Value: &pgtype.OIDValue{},
-						Name:  "int8OID",
-						OID:   pgtype.Int8OID,
-					})
-					// Newer versions of pgbouncer need this defined. See the discussion here:
-					// https://github.com/jackc/pgx/issues/649
-					info.RegisterDataType(pgtype.DataType{
-						Value: &pgtype.OIDValue{},
-						Name:  "numericOID",
-						OID:   pgtype.NumericOID,
-					})
-
-					return info, nil
-				},
-			},
```
It does look like this block is important to pgbouncer support.  Does pgx v4 handle this in another way or does this PR remove pgbouncer support?

ZPascal replied:
In my view, the configuration of the driver should be automatic and the manual setup is no longer needed in v4. Furthermore, the PR should only fix a bug and not remove a feature. To eliminate this risk I will first try to create an integration test.

ZPascal commented:
Hi @helenosheaa,

I have looked at the PGBouncer integration test and run it in the main and feature branches. Here I could see the same error message in both branches. I get the following error message and the test is normally skipped:

**Container:**
`failed to connect to `host=localhost user=pgbouncer database=pgbouncer`: server error (ERROR: not allowed (SQLSTATE 08P01)).`

**Console:**
```
2021-05-03 14:41:11.165 UTC [1] LOG C-0x7f028b739100: pgbouncer/pgbouncer@172.21.0.1:61820 login attempt: db=pgbouncer user=pgbouncer tls=no
2021-05-03 14:41:11.165 UTC [1] LOG C-0x7f028b739100: pgbouncer/pgbouncer@172.21.0.1:61820 closing because: not allowed (age=0s)
2021-05-03 14:41:11.165 UTC [1] WARNING C-0x7f028b739100: pgbouncer/pgbouncer@172.21.0.1:61820 pooler error: not allowed
```

Furthermore, I can also see a corresponding error message in both scenarios in the container log. I am not a PGBouncer expert and would like to know if there is anything I need to consider when setting up the containers (PostgreSQL + PGBouncer) and running the tests? When was the last time the tests were run?

Anyway, I then looked at the PgBouncer integration test and the configuration of the connection within service.go and decided to use the PgBouncer integration test. I subsequently modified the container used accordingly and published it to Dockerhub. Furthermore, I modified the service.go according to the issue linked [here](https://github.com/jackc/pgx/issues/650). PgBouncer does not know the configuration parameter ```statement_cache_mode=describe```.

Afterwards I modified the integration test of PgBouncer, because there seems to be an update in PgBouncer and now other metrics are returned in a modified structure.

**Summary**
The adaptation of the integration test, the update of the container and the configuration of the connection allow the further use of PgBouncer and the execution of the PgBouncer integration test without having to skip it.

ssoroka commented on scripts/check-deps.sh:
```diff
@@ -28,8 +28,8 @@ for dep in $(LC_ALL=C sort -u "${tmpdir}/golist"); do
-		# dependency is replaced in go.mod
-		github.com/satori/go.uuid) continue;;
+    # dependency is replaced in go.mod
+    github.com/satori/go.uuid) continue;;
```
can you restore the original spacing here? Thanks!

ZPascal replied:
Thanks for the hint. It's done.

helenosheaa commented:
@ZPascal thanks for working on the integration test! I don't think it was in a working state previously so I really appreciate it.

What was the reason for changing to your own docker image?

I saw that you added `d.PreferSimpleProtocol = true` which was in the previous config. It may turn out we need to add in more custom config but if you address the merge conflicts, I'm happy to approve it as you've got the integration test working.

ZPascal commented:
@helenosheaa Gladly, no problem

> What was the reason for changing to your own docker image?

I had seen that the existing image was no longer regularly maintained and updated. Furthermore, the stats_users parameter must be set in the corresponding [configuration](https://github.com/ZPascal/pgbouncer-container/blob/main/run.sh#L30). For this reason, I decided to create an updated image.

> I saw that you added d.PreferSimpleProtocol = true which was in the previous config.
Right. It's necessary to set the parameter `PreferSimpleProtocol = true` at least in the case of PgBouncer. It's documented inside the following [issue](https://github.com/jackc/pgx/issues/650).

helenosheaa reviewed (approved):
Ok, thanks for clarifying! 

@reimda are you also happy for this to be merged now there's a working integration test?

ZPascal commented:
@helenosheaa Any news ?

ssoroka commented on plugins/inputs/postgresql/service.go:
```diff
@@ -111,33 +110,16 @@ func (p *Service) Start(telegraf.Accumulator) (err error) {
-				},
-				CustomConnInfo: func(c *pgx.Conn) (*pgtype.ConnInfo, error) {
-					info := c.ConnInfo.DeepCopy()
-					info.RegisterDataType(pgtype.DataType{
-						Value: &pgtype.OIDValue{},
-						Name:  "int8OID",
-						OID:   pgtype.Int8OID,
-					})
-					// Newer versions of pgbouncer need this defined. See the discussion here:
-					// https://github.com/jackc/pgx/issues/649
-					info.RegisterDataType(pgtype.DataType{
-						Value: &pgtype.OIDValue{},
-						Name:  "numericOID",
-						OID:   pgtype.NumericOID,
-					})
```
it looks like we're still losing support for these data types? Do you know that they're both covered by the new v4 library?

ZPascal replied:
The mapping of the [data types](https://github.com/jackc/pgtype/blob/b26cd223783bf410ec9cd789ae872cef56610d93/pgtype.go#L65) is done inside the driver. Within pgx v4 the mapping has been outsourced and there is a separate repository for it. The version of the driver I am using, uses 1.3.0 of the outsourced types repositories. Furthermore, I have also made the corresponding integration test functional and could already test the corresponding extraction of the metrics there.

ssoroka replied:
got it. just trying to make sure we didn't inadvertently remove support for these two types. Otherwise we're good to go

reimda replied:
Thanks for the link to the pgtype mapping. I didn't know that's how it works but it makes sense now.
